src/processor/graph.monitor.processor.service.ts

```diff
@@ -0,0 +1,137 @@
+import { InjectRedis } from '@liaoliaots/nestjs-redis';
+import { InjectQueue, Processor, WorkerHost } from '@nestjs/bullmq';
+import { Injectable, Logger } from '@nestjs/common';
+import { Job, Queue } from 'bullmq';
+import Redis from 'ioredis';
+import { MILLISECONDS_PER_SECOND } from 'time-constants';
+import { RegistryError } from '@polkadot/types/types';
+import { BlockchainService } from '#app/blockchain/blockchain.service';
+import { BlockchainConstants } from '#app/blockchain/blockchain-constants';
+import { ITxMonitorJob } from '#app/interfaces/monitor.job.interface';
+import { IGraphUpdateJob } from '#app/interfaces/graph-update-job.interface';
+
+@Injectable()
+@Processor('graphTxMonitorQueue')
+export class GraphNotifierService extends WorkerHost {
+  private logger: Logger;
+
+  constructor(
+    @InjectRedis() private cacheManager: Redis,
+    @InjectQueue('graphUpdateQueue') private reconnectionQueue: Queue,
+    private blockchainService: BlockchainService,
+  ) {
+    super();
+    this.logger = new Logger(this.constructor.name);
+  }
+
+  async process(job: Job<ITxMonitorJob, any, string>): Promise<any> {
+    this.logger.log(`Processing job ${job.id} of type ${job.name}`);
+    try {
+      const numberBlocksToParse = BlockchainConstants.NUMBER_BLOCKS_TO_CRAWL;
+      const txCapacityEpoch = job.data.epoch;
+      const previousKnownBlockNumber = (await this.blockchainService.getBlock(job.data.lastFinalizedBlockHash)).block.header.number.toBigInt();
+      const currentFinalizedBlockNumber = await this.blockchainService.getLatestFinalizedBlockNumber();
+      const blockList: bigint[] = [];
+
+      for (let i = previousKnownBlockNumber; i <= currentFinalizedBlockNumber && i < previousKnownBlockNumber + numberBlocksToParse; i += 1n) {
+        blockList.push(i);
+      }
+      const txResult = await this.blockchainService.crawlBlockListForTx(job.data.txHash, blockList, [{ pallet: 'system', event: 'ExtrinsicSuccess' }]);
+      if (!txResult.found) {
+        this.logger.error(`Tx ${job.data.txHash} not found in block list`);
+        throw new Error(`Tx ${job.data.txHash} not found in block list`);
+      } else {
+        // Set current epoch capacity
+        await this.setEpochCapacity(txCapacityEpoch, BigInt(txResult.capacityWithDrawn ?? 0n));
```

Comment:
The epoch when the tx was submitted may not be the same epoch that it was included in a block. We want to record the epoch it actually used capacity in, not necessarily when it was submitted.

Reply:
I think that would rarely be different around the edge of epoch, essentially its just for checking if we have capacity or not and have provision to pause or handle low capacity error if at all they happen in a production setting for a provider

Reply:
Still, for correctness, I think it would make sense to record the epoch in which the capacity was actually withdrawn.